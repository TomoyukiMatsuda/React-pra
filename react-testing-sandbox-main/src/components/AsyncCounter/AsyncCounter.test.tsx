import { act, cleanup, fireEvent, render, screen } from "@testing-library/react"

import { AsyncCounter } from "./"

describe("AsyncCounter", () => {
  afterEach(() => {
    cleanup()
  })

  test("render", () => {
    const { asFragment } = render(<AsyncCounter />)
    expect(asFragment()).toMatchSnapshot()
  })

  describe("click:count:カウントアップ", () => {
    test("ボタン押下 1 秒後は 1 カウントアップ", () => {
      jest.useFakeTimers() /** 時間詐称 */
      render(<AsyncCounter />)
      const button = screen.getByText("AsyncIncrement")
      fireEvent.click(button)
      act(() => {
        jest.runAllTimers()
      })
      screen.getByText("AsyncCount: 1")
      jest.useRealTimers() /** 時を戻そう */
    })
  })

  describe("click:count:ボタン活性・非活性", () => {
    test("ボタン押下直後はボタンが非活性", () => {
      render(<AsyncCounter />)
      const button = screen.getByText("AsyncIncrement")
      fireEvent.click(button)
      expect(button).toBeDisabled()
    })
    test("ボタン押下 1 秒後はボタンが活性", () => {
      jest.useFakeTimers() /** 時間詐称 */
      render(<AsyncCounter />)
      const button = screen.getByText("AsyncIncrement")
      fireEvent.click(button)
      act(() => {
        jest.runAllTimers()
      })
      expect(button).not.toBeDisabled()
      jest.useRealTimers() /** 時を戻そう */
    })
  })

  describe("click:count:ローディング UI", () => {
    test("ボタン押下直後はローディングが表示", () => {
      render(<AsyncCounter />)
      const button = screen.getByText("AsyncIncrement")
      fireEvent.click(button)
      expect(screen.queryByText("...Loading")).toBeInTheDocument()
    })
    test("ボタン押下直後はローディングが非表示", () => {
      jest.useFakeTimers() /** 時間詐称 */
      render(<AsyncCounter />)
      const button = screen.getByText("AsyncIncrement")
      fireEvent.click(button)
      act(() => {
        jest.runAllTimers()
      })
      expect(screen.queryByText("...Loading")).not.toBeInTheDocument()
      jest.useRealTimers() /** 時を戻そう */
    })
  })
})
