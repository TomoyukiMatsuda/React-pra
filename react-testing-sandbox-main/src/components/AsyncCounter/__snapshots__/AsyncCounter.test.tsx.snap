// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncCounter render 1`] = `
<DocumentFragment>
  <div>
    AsyncCount: 0
  </div>
  <div>
    <button>
      AsyncIncrement
    </button>
  </div>
</DocumentFragment>
`;
